hudi-client/src/main/java/org/apache/hudi/metadata/HoodieMetadataImpl.java

```diff
@@ -0,0 +1,1003 @@
+              timeline.getInstantDetails(instant).get(), HoodieCommitMetadata.class);
+          update(commitMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.ROLLBACK_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieRollbackMetadata rollbackMetadata = TimelineMetadataUtils.deserializeHoodieRollbackMetadata(
+              timeline.getInstantDetails(instant).get());
+          update(rollbackMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.RESTORE_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          HoodieRestoreMetadata restoreMetadata = TimelineMetadataUtils.deserializeHoodieRestoreMetadata(
+              timeline.getInstantDetails(instant).get());
+          update(restoreMetadata, instant.getTimestamp());
+          break;
+        }
+        case HoodieTimeline.SAVEPOINT_ACTION: {
+          ValidationUtils.checkArgument(instant.isCompleted());
+          // Nothing to be done here
+          break;
+        }
+        default: {
+          throw new HoodieException("Unknown type of action " + instant.getAction());
+        }
+      }
+    }
+  }
+
+  /**
+   * Return an ordered list of instants which have not been synced to the Metadata Table.
+
+   * @param datasetMetaClient {@code HoodieTableMetaClient} for the dataset
+   */
+  public List<HoodieInstant> findInstantsToSync(HoodieTableMetaClient datasetMetaClient) {
+    HoodieTimeline metaTimeline = metaClient.reloadActiveTimeline();
+
+    // All instants since the last time metadata table was compacted are candidates for sync
+    Option<String> compactionTimestamp = getLatestCompactionTimestamp();
+
+    // If there has not been any compaction then the first delta commit instant should be the one at which
+    // the metadata table was created. We should not sync any instants before that creation time.
+    Option<HoodieInstant> oldestMetaInstant = Option.empty();
+    if (!compactionTimestamp.isPresent()) {
+      oldestMetaInstant = Option.fromJavaOptional(metaTimeline.getInstants()
+          .filter(i -> i.isCompleted() && i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION))
+          .findFirst());
+      if (oldestMetaInstant.isPresent()) {
+        // TODO: Ensure this is the instant at which we created the metadata table
+      }
+    }
+
+    String metaSyncTimestamp = compactionTimestamp.isPresent() ? compactionTimestamp.get()
+        : oldestMetaInstant.isPresent() ? oldestMetaInstant.get().getTimestamp() : "";
+
+    // Metadata table is updated when an instant is completed except for the following:
+    //  CLEAN: metadata table is updated during inflight. So for CLEAN we accept inflight actions.
+    List<HoodieInstant> datasetInstants = datasetMetaClient.getActiveTimeline().getInstants()
+        .filter(i -> i.isCompleted() || (i.getAction().equals(HoodieTimeline.CLEAN_ACTION) && i.isInflight()))
```

Comment:
but when we are syncing from timeline. its better to not rely on inflight instants right? for e.g what if there is a .inflight clean that never does completed and the user redeployed with a different config for cleaning, which results in a different clean plan the next time.

Reply:
Correction. I think we will still honor scheduled clean plans. but overall, the point remains the same. we can just reflect completed cleans IMO. The only side effect would be that when cleaning by number of retained versions, it can lead to less aggressive cleaning in rare cases. for cleaning by commits, it should be fine, as the cleaning is decided based on a watermark commit time  and file slices before that are discarded.

Reply:
See my note on reflecting compaction after its completed. It has some issues for queries.